README.md
# MapBBCode Share

This is a website originally made for sharing maps, based on [MapBBCode](https://github.com/mapbbcode/mapbbcode) javascript library. But since it has powerful import/export feature, it also can be used for planning routes, generating standalone map pages and converting between formats.

## Installation

In order to reduce load on the main [share.mapbbcode.org](http://share.mapbbcode.org) website, cloning the service is encouraged. Requirements are few:

* Apache with mod_rewrite or nginx with php-fpm
* PHP 5.2 or newer
* MySQL 5
* A separate domain (or subdomain) name

To install, copy all files (except `README.md`) to a designated directory, modify `config.php` according to your setup, and open `http://domain.name/initdb`. If you see a map with a yellow message saying tables were created, you're good to go. Set `NEED_INIT_DB` in the config to `false`, then check that saving and signing in work.

For caching to work, you need a `cache` directory with writing rights.

## Formats

At the moment MapBBCode Share can import 11 file types and export 10. Writing a plugin for a new file format is easy: see examples in `formats` directory. Those are loaded automatically, you won't have to modify other scripts. Keep in mind this is not a universal convertor: [GPSBabel](http://www.gpsbabel.org/) is. This is not a GPS trace storage, and traces are simplified: we already have [GPSies](http://www.gpsies.com/) for full GPX downloads, elevation and speed profiles. MapBBCode Share goals are simple and outlined in the first section, and supported formats list should be kept short.

## Included Libraries

Writing the service would be much harder it it weren't for those libraries:

* [MapBBCode](https://github.com/mapbbcode/mapbbcode) (WTFPL)
* [Leaflet](http://leafletjs.com/) (BSD 2-clause)
* [Leaflet.draw](https://github.com/Leaflet/Leaflet.draw) (MIT)
* [Bing.js](https://github.com/shramov/leaflet-plugins/blob/master/layer/tile/Bing.js) (BSD 2-clause)
* [LightOpenID](http://code.google.com/p/lightopenid/) (MIT)
* [OpenID Selector](http://code.google.com/p/openid-selector/) (BSD 3-clause)
* [Simplify.js](http://mourner.github.io/simplify-js/) (BSD 2-clause)
* [Simplify.php](https://github.com/AKeN/simplify-php) (unknown license)

## License

All files that did not come from other libraries are released under [WTFPL](http://www.wtfpl.net/).


CHANGELOG.md
# MapBBCode Share Changelog

Versions are numbered `N.M-S`, where `N.M` is the MapBBCode library version, and `S` is the Share build, independent of the library version. There are no "stable" versions and releases: the latest version is always the one to install.

### 1.2-15, 2018-06-26

* Major update to CSV parsing module, detects more formats now.

### 1.2-14, 2015-04-06

* Option to remove maps from libraries. [#5](https://github.com/MapBBCode/share.mapbbcode.org/pull/5)
* A button in the library pane to merge accounts. [#43](https://github.com/MapBBCode/share.mapbbcode.org/pull/43)
* More obvious editing link. [#24](https://github.com/MapBBCode/share.mapbbcode.org/pull/24)
* Made iframe option visible under "Embed" link. [#41](https://github.com/MapBBCode/share.mapbbcode.org/pull/41)
* A warning message about Google retiring its OpenID.

### 1.2-13, 2014-03-06

* Fix processing of delimiters inside quotes for CSV.

### 1.2-12, 2013-12-28

* Move default position, zoom and layers to `config.php` (by [@aaronpk](https://github.com/aaronpk)). [#23](https://github.com/MapBBCode/share.mapbbcode.org/pull/23)
* Support for nginx with php-fpm (by [@aaronpk](https://github.com/aaronpk)). [#20](https://github.com/MapBBCode/share.mapbbcode.org/pull/20)
* Swap MyOpenID for Vkontakte OpenID provider. [#21](https://github.com/MapBBCode/share.mapbbcode.org/issues/21)
* Fix paths to `auth.php` for cases when Share is not in a root directory.

### 1.2-11, 2013-12-27

* `?direct` parameter does not add attachment headers (useful for HTML export).
* Updated L.PopupIcon in HTML format exporter.
* Host detection in `auth.php` instead of a hardcoded address. [#15](https://github.com/MapBBCode/share.mapbbcode.org/issues/15)
* IFrame format. [#17](https://github.com/MapBBCode/share.mapbbcode.org/issues/17)

### 1.2-10, 2013-12-20

* Updated MapBBCode to 1.2.0-dev.
* Process colors in GPX.

### 1.1-9, 2013-12-12

* Fixed exporting empty maps and single marker maps to HTML. [#10](https://github.com/MapBBCode/share.mapbbcode.org/issues/10)
* Exported HTML now uses Leaflet 0.7.1. [#11](https://github.com/MapBBCode/share.mapbbcode.org/issues/11)
* WPT import uses description field if it's not empty. [#12](https://github.com/MapBBCode/share.mapbbcode.org/issues/12)
* Quotes in WPT and PLT are now screened, and are read properly.
* Export tries to create filename out of map title before resorting to code id (sadly, cyrillic letters are dropped).
* Increased displayed library size to 100 maps.

### 1.1-8, 2013-12-01

* Updated MapBBCode to 1.1.2.
* Extracted help contents to `help.txt`.

### 1.1-7, 2013-11.29

* Updated MapBBCode to 1.1.2-dev.
* Added attribution edit link
* Fixed cache purging.

### 1.1-6, 2013-11-26

* Updated MapBBCode to 1.1.2-dev.
* Added length measurement with a new plugin.

### 1.1-5, 2013-11-23

* Website now works in subfolders and without mod_rewrite. [#4](https://github.com/MapBBCode/share.mapbbcode.org/issues/4)
* Fixed edit id reset after importing a file. [#7](https://github.com/MapBBCode/share.mapbbcode.org/issues/7)
* OziExplorer formats now can have their own encoding (usually iso or cp1251). [#6](https://github.com/MapBBCode/share.mapbbcode.org/issues/6)

### 1.1-4, 2013-11-16

* Updated MapBBCode to version 1.1.1.

### 1.1-3, 2013-11-13

* Updated MapBBCode to version 1.1.0.
* Updated MapBBCode parsing functions, semicolons in titles are now processed correctly.
* Forgot to replace one bbcode specification link.

### 1.0-2, 2013-11-12

* Empty CSV now doesn't have a header to not confuse users; single path is saved as csv.
* BBCode tags can be omitted in `?bbcode=` parameter.
* Moved Bing key to a configuration file.
* Ability to work without a database. [#3](https://github.com/MapBBCode/share.mapbbcode.org/issues/3)
* Merge imported traces with existing data. [#2](https://github.com/MapBBCode/share.mapbbcode.org/issues/2)
* Highlight "Save" button on map change. [#1](https://github.com/MapBBCode/share.mapbbcode.org/issues/1)
* Rephrased the message with a link for sharing.
* New OpenMapSurfer tile URL.
* New configuration keys: `BING_KEY`, `IMPORT_SINGLE`.

### 1.0-1, 2013-11-01

* Added Cycle Map layer.
* Rephrased the message with a link for sharing.
* Restricted bots from editing pages.

### 1.0-0, 2013-10-31

Initial release.


index.php
<?php
define('IN_MAPBBCODE', 1);
define('VERSION', '1.2-15');
require('config.php');
require('convert.php');
require('db.php');

define('SCRIPT_NAME', 'index.php');
if( !defined('MOD_REWRITE') ) {
	if( function_exists('apache_get_modules') )
		define('MOD_REWRITE', in_array('mod_rewrite', apache_get_modules()) && file_exists('.htaccess'));
	else
		define('MOD_REWRITE', true); // true for nginx
}

$doc_path = 'http'.(!empty($_SERVER['HTTPS']) && $_SERVER['HTTPS'] !== 'off' ? 's' : '').'://'.
    $_SERVER['HTTP_HOST'].rtrim(dirname($_SERVER['SCRIPT_NAME']), '/\\');
$base_path = MOD_REWRITE ? $doc_path : $doc_path.'/'.SCRIPT_NAME;

ini_set('session.gc_maxlifetime', 7776000);
ini_set('session.cookie_lifetime', 7776000);
session_set_cookie_params(7776000);
session_start();

//if( $_SERVER['HTTP_HOST'] == 'localhost' ) $userid = 'test';

if( isset($_SESSION['user_id']) ) {
    $userid = $_SESSION['user_id'];
    if( isset($_SESSION['alt_user_id']) ) {
        check_alt_user_id($userid, $_SESSION['alt_user_id']);
        unset($_SESSION['alt_user_id']);
    }
}

$message = '';
$api = isset($_REQUEST['api']);
$params = parse_params();
$action = $params['action'];
$bbcode = $params['bbcode'];
$title = $params['title'];

if( isset($params['id']) )
    $scodeid = $params['id'];
if( isset($params['editid']) )
    $seditid = $params['editid'];

$data = false;
if( isset($scodeid) ) {
    $data = get_data($scodeid);
    if( !$data ) {
        if( $api )
            return_json(array('error' => 'No such code: '.$scodeid));
        $message = 'There is no code in the database with given id';
    }
}

if( $action == '' && $data && !$params['post'] ) {
    // only for GET requests with id (that is, not ?bbcode=)
    $bbcode = $data['bbcode'];
    $title = $data['title'];

    if( $api )
        return_json(array('title' => $title, 'bbcode' => $bbcode));

} elseif( $action == '' && isset($_GET['gz']) && strlen($_GET['gz']) > 12 ) {
    // action is base64-encoded bbcode+title
    $res = decompress_bbcode($_GET['gz']);
    if( $res && count($res) == 2 && strlen($res[1]) > 10 ) {
        $title = $res[0];
        $bbcode = $res[1];
    }
}

if( $data && isset($params['key']) && $params['key'] == $data['editid'] ) {
    $newcode = false;
    $seditid = $params['key'];
    $message = '<b><a href="'.$base_path.'/'.$params['id'].'" target="mapbbstatic">Share this link</a></b> for read-only view of this map.<br><a href="'.$base_path.'/'.$params['id'].'?editid='.$params['key'].'">Bookmark this</a> to edit the map later';
    $nohide = 1; // do not hide message
    if( isset($userid) ) {
        update_library($userid, $scodeid, $seditid);
        $message .= ' (or check the library)';
    } elseif( db_available() )
        $message .= ' (sign in to have it stored for you automatically)';
}

$editing = $params['post'] || isset($seditid) || (strlen($bbcode) == 0 && strlen($title) == 0);
$newcode = !isset($seditid);

// see also $actions array in parse_surl
if( $action == 'initdb' && NEED_INIT_DB ) {
    initdb();

} elseif( $action == 'fmtlist' ) {
    $fmtdesc = get_format_arrays();
    return_json($fmtdesc);

} elseif( $action == 'save' ) {
    save($params, $data);

} elseif( $action == 'forget' && isset($userid) && isset($scodeid) ) {
    if( remove_bookmark($userid, $scodeid) )
        $message .= 'This map has been removed from your library.';

} elseif( $action == 'signout' ) {
    unset($_SESSION['user_id']);
    unset($userid);

} elseif( $action == 'bookmark' && isset($userid) && $data ) {
    update_library($userid, $params['id'], '');
    $message = 'The code was added to your library';

} elseif( $action == 'import' && isset($_FILES['file']) ) {
    if( !is_uploaded_file($_FILES['file']['tmp_name']) || $_FILES['file']['size'] == 0 || $_FILES['file']['error'] > 0 ) {
        $errors = array('OK', 'too big', 'bigger than MAX_FILE_SIZE', 'partial upload', 'no file', '', 'nowhere to store', 'failed to write', 'extension error');
        $err = $errors[$_FILES['file']['error']];
        if( $api )
            return_json(array('error' => 'File upload error: '.$err, 'code' => $_FILES['file']['error']));
        else
            $message = 'File upload error ('.$err.')';
    } else {
        $titlebb = defined('IMPORT_SINGLE') && !IMPORT_SINGLE ? array($title, $bbcode) : array('', '');
        $titlebb = import($_FILES['file']['tmp_name'], $titlebb);
        $title = $titlebb[0];
        $bbcode = $titlebb[1];
        if( $api )
            return_json(array('title' => $title, 'bbcode' => $bbcode));
    }
}

if( isset($_REQUEST['format']) && preg_match('/^[a-z]+$/', $_REQUEST['format']) ) {
    $format = $_REQUEST['format'];
    header('Access-Control-Allow-Origin: *');
    $result = export($format, $title, $bbcode, isset($scodeid) ? $scodeid : '', !isset($_REQUEST['direct']));
    if( $result == CONVERT_OK )
        exit;
    if( $result == CONVERT_NOT_SUPPORTED ) {
        if( $action == 'export' )
            $message = 'Unknown or unimplemented export type: '.$format;
        else {
            header('HTTP/1.1 415 Format Not Suppoprted');
            exit;
        }
    } elseif( $result == CONVERT_EMPTY ) {
        if( $action == 'export' )
            $message = 'Output file is empty';
        else {
            header('HTTP/1.1 400 No Data In BBCode');
            exit;
        }
    }
}

$fmtdesc = get_format_arrays();

if( !$editing && strlen($bbcode) < 11 )
    $bbcode = '[map][/map]';

if( isset($userid) )
    $library = fetch_library($userid);

if( isset($userid) && !isset($scodeid) )
    $message .= 'Warning: Google retires its OpenID on April 20th, merge your account with another one to keep your library.';

require('page.php');

// ---------------------------------- FUNCTIONS ----------------------------

// print data as json(p) and exit
function return_json( $data ) {
    header('Access-Control-Allow-Origin: *');
    header('Content-Type: application/json');
    $json = json_encode($data);
    if( isset($_REQUEST['jsonp']) )
        print $_REQUEST['jsonp']."($json);";
    else
        print $json;
    exit;
}

// save bbcode to database (or update)
function save( $params, $data ) {
    global $message, $api, $base_path;
    $title = $params['title'];
    $bbcode = $params['bbcode'];

    if( strlen($title) < 3 && strlen($bbcode) < 7 ) {
        if( $api )
            return_json(array('error' => 'Would not save empty data, sorry'));
        else
            $message = 'Would not save empty data, sorry';
        return;
    }

    if( strlen($title) > 250 ) {
        $spacepos = strrpos($title, ' ');
        $title = substr($title, 0, $spacepos !== false ? $spacepos : 250);
    }

    $db = getdb();
    if( $data && isset($params['key']) && $data['editid'] == $params['key'] ) {
        // update
        $codeid = $params['id'];
        $editid = $params['key'];
        $sql = !$db ? '' : "update ".DB_TABLE." set updated=now(), title='".$db->escape_string($title)."', bbcode='".$db->escape_string($bbcode)."' where codeid = '$codeid'";
        cache_remove($codeid, 'code');
        cache_remove(false, 'user'); // yup, now a lot of users can have their libraries updated
    } else {
        $editid = generate_id(EDIT_HASH_LENGTH);
        $tries = 10;
        do {
            $codeid = generate_id(HASH_LENGTH);
            $exists = get_data($codeid) !== false;
        } while( $exists );
        $sql = !$db ? '' : "insert into ".DB_TABLE." (created, updated, codeid, editid, title, bbcode) values(now(), now(), '$codeid', '$editid', '".$db->escape_string($title)."', '".$db->escape_string($bbcode)."')";
    }

    if( $db ) {
        $res = $db->query($sql);
    } else {
        // put code to cache
        $assoc = array('editid' => $editid, 'title' => $title, 'bbcode' => $bbcode);
        cache_put($codeid, 'code', $assoc);
        $res = true;
    }
    if( !$api ) {
        if( !$res ) {
            $message = 'Failed to insert entry in the database: '.$db->error;
        } else {
            header("Location: ".$base_path."/$codeid?editid=$editid");
            exit;
        }
    } else {
        if( !$res ) {
            return_json(array('error' => 'Failed to insert entry in the database: '.$db->error));
        } else {
            return_json(array('codeid' => $codeid, 'editid' => $editid, 'viewurl' => $base_path."/$codeid", 'editurl' => $base_path."/$codeid/$editid"));
        }
        exit;
    }
}

// Adds code id to the user's library. Edit id is checked to test if user can edit code
function update_library( $userid, $codeid, $editid ) {
    $db = getdb();
    if( !$db )
        return;
    $uidesc = $db->escape_string($userid);
    $sql = 'select editable from '.DB_TABLE."_users where codeid = '$codeid' and userid = '$uidesc'";
    $res = $db->query($sql);
    if( !$res )
        return;
    $cnt = $res->num_rows;
    $row = $res->fetch_row();
    $editable = $row[0];
    $res->free();
    if( !$cnt ) {
        $sql = 'insert into '.DB_TABLE."_users (userid, codeid, editable) values ('$uidesc', '$codeid', ".(isset($editid) && strlen($editid) > 0 ? 1 : 0).")";
        $db->query($sql);
    } elseif( !$editable && isset($editid) && strlen($editid) > 0 ) {
        $sql = 'update '.DB_TABLE."_users set editable=1 where userid='$uidesc' and codeid='$codeid'";
        $db->query($sql);
    }
    cache_remove($userid, 'user');
}

// Removes a bookmark from user's library
function remove_bookmark( $userid, $codeid ) {
    $db = getdb();
    if( !$db )
        return false;
    $sql = 'delete from '.DB_TABLE."_users where codeid = '$codeid' and userid ".get_alt_userid_clause($db, $userid);
    $res = $db->query($sql);
    if( $res && $db->affected_rows > 0 ) {
        cache_remove($userid, 'user');
        return true;
    }
    return false;
}

// Merges two user accounts
function check_alt_user_id( $userid, $alt_userid ) {
    $db = getdb();
    if( !$db )
        return;
    $uidesc = $db->escape_string($userid);
    $alt_uidesc = $db->escape_string($alt_userid);
    $sql = 'select * from '.DB_TABLE."_usermap where (userid = '$uidesc' and puserid = '$alt_uidesc') or (userid = '$alt_uidesc' and puserid = '$uidesc')";
    $res = $db->query($sql);
    if( !$res )
        return;
    if( !$res->num_rows ) {
        $sql = 'insert into '.DB_TABLE."_usermap (userid, puserid) values ('$alt_uidesc', '$uidesc')";
        $db->query($sql);
    }
    cache_remove($userid, 'user');
    cache_remove($alt_userid, 'user');
}

// Updates user clause for merged accounts
function get_alt_userid_clause( $db, $userid ) {
    $esc_userid = $db->escape_string($userid);
    $sql = 'select userid from '.DB_TABLE."_usermap where puserid = '$esc_userid' union select puserid as userid from ".DB_TABLE."_usermap where userid = '$esc_userid'";
    $res = $db->query($sql);
    $ids = array("'$esc_userid'");
    if( $res )
        while( $row = $res->fetch_row() )
            $ids[] = "'".$db->escape_string($row[0])."'";
    if( count($ids) == 1 )
        return '= '.$ids[0];
    else
        return 'in ('.implode(',', $ids).')';
}

// Returns as array of all library entries sorted by update time
function fetch_library( $userid ) {
    $stored = cache_fetch($userid, 'user');
    if( $stored !== false )
        return $stored;

    global $message;
    $codes = array();
    $db = getdb();
    if( !$db )
        return $codes;
    $user_clause = get_alt_userid_clause($db, $userid);
    $sql = 'select now() as now, m.*, u.editable from '.DB_TABLE.' m, '.DB_TABLE.'_users u where u.codeid = m.codeid and u.userid '.$user_clause.' order by m.updated desc limit 100';
    $res = $db->query($sql);
    if( $res ) {
        require_once('mapbbcode.php');
        while( $row = $res->fetch_assoc() ) {
            $item = array();
            $item['codeid'] = $row['codeid'];
            $item['editid'] = $row['editable'] ? $row['editid'] : '';
            $item['created'] = human_date($row['created'], $row['now']);
            $item['updated'] = human_date($row['updated'], $row['now']);
            $item['title'] = $row['title'];
            $item['stats'] = mapbbcode_stats($row['bbcode']);
            $codes[] = $item;
        }
        $res->free();
        cache_put($userid, 'user', $codes);
    } else
        $message = 'Failed to retrieve user library: '.$db->error;
    return $codes;
}

// Generated random id with given length
function generate_id($length) {
    $id = '';
    for( $i = 0; $i < $length; $i++ )
        $id .= chr(mt_rand(97, 122)); // 'a'..'z'
    return $id;
}

// Arhives bbcode and title
function compress_bbcode($title, $bbcode) {
    $base64 = base64_encode(gzdeflate($bbcode.str_replace('[/map]', '[ /map]', $title), 9));
    $firsteq = strpos($base64, '=');
    return strtr($firsteq === false ? $base64 : substr($base64, 0, $firsteq), '+/', '-_');
}

// Unarchives bbcode and title
function decompress_bbcode($str) {
    $base64 = strtr($str, '-_', '+/');
    while( strlen($base64) % 4 )
        $base64 .= '=';
    $titlebb = gzinflate(base64_decode($base64), 20000);
    if( $titlebb === false )
        return false;
    $bbpos = strpos($titlebb, '[/map]');
    if( $bbpos === false ) {
        $title = $titlebb;
        $bbcode = '';
    } else {
        $title = substr($titlebb, $bbpos + 6);
        $bbcode = substr($titlebb, 0, $bbpos + 6);
    }
    return array(str_replace('[ /map]', '[/map]', $title), $bbcode);
}

// Screens \ and ' characters (for javascript strings)
function screen_param($str) {
    return str_replace(array("\\", "'"), array("\\\\", "\\'"), $str);
}

// Converts mysql date to human-readable string
function human_date($sqldate, $sqlnow = false) {
    $now = $sqlnow ? strtotime($sqlnow) : time();
    $date = strtotime($sqldate);
    $diff = round(($now - $date) / 60); // in minutes
    //echo "($sqldate) ".date('Y-m-d H:i:s', $date).' -> '.date('Y-m-d H:i:s', $now)." = $diff<br>";
    if( $diff < 0 )
        return 'in the future: '.$diff;
    if( $diff == 0 ) return 'just now';
    if( $diff == 1 ) return 'a minute ago';
    if( $diff < 55 ) return $diff.' minutes ago';
    $hours = round($diff/60);
    $curhour = (int)date('H', $now);
    if( $hours == 1 ) return 'an hour ago';
    if( $hours <= max(12, $curhour + 1) ) return $hours.' hours ago';
    // now test real difference in dates
    $curday = floor(strtotime(date('Y-m-d', $now)) / 86400);
    $day = floor(strtotime(date('Y-m-d', $date)) / 86400);
    if( $curday - $day == 1 ) return 'yesterday';
    // if( $curday - $day == 2 ) return 'two days ago';
    // now output real dates
    //$months = array('', 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'June', 'July', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec');
    $months = array('', 'January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December', '');
    $month = $months[(int)idate('m', $date)];
    if( date('Y', $now) == date('Y', $date) )
        return $month.' '.idate('d', $date);
    else
        return $month.' '.date('Y', $date);
}

// ---------------------------------- URL PARSING ----------------------------

// returns URL for those three parameters
function build_url( $action, $codeid = '', $editid = '' ) {
    global $base_path;
    $url = $base_path;
    if( strlen($codeid) > 0 ) {
        $url .= '/'.$codeid;
        if( strlen($editid) > 0 )
            $url .= '/'.$editid;
    }
    if( strlen($action) > 0 )
        $url .= '?'.$action;
    return $url;
}

// returns array(action, id?, key?)
function parse_surl() {
    $base_path = rtrim(dirname($_SERVER['SCRIPT_NAME']), '/\\');

    if(function_exists('apache_request_headers'))
        $url = isset($_SERVER['REDIRECT_URL']) ? $_SERVER['REDIRECT_URL'] : $_SERVER['PHP_SELF'];
    else
        $url = $_SERVER['REQUEST_URI']; // under nginx, path info is in REQUEST_URI

    if( substr($url, 0, strlen($base_path)) == $base_path )
        $url = substr($url, strlen($base_path));
    if( substr($url, 1, strlen(SCRIPT_NAME)) == SCRIPT_NAME )
        $url = substr($url, strlen(SCRIPT_NAME) + 1);

    $result = array('action' => '');
    $actions = array('initdb', 'fmtlist', 'save', 'signout', 'bookmark', 'import', 'export', 'forget');
    if( preg_match('#^/?([a-z]+)(?:/([a-z]+))?/?#', $url, $m) ) {
        $result[in_array($m[1], $actions) ? 'action' : 'id'] = $m[1];
        $result['key'] = count($m) > 2 ? $m[2] : '';
    }
    return $result;
}

// returns array(action, id?, key?, codeid?, editid?, title, bbcode)
function parse_params() {
    $result = parse_surl();
    if( isset($_REQUEST['codeid']) && preg_match('/^[a-z]+$/', $_REQUEST['codeid']) && strlen($_REQUEST['codeid']) == HASH_LENGTH ) {
        $result['codeid'] = $_REQUEST['codeid'];
        $result['id'] = $result['codeid']; // POST overrides GET
    } if( isset($_REQUEST['editid']) && preg_match('/^[a-z]+$/', $_REQUEST['editid']) ) {
        $result['editid'] = $_REQUEST['editid'];
        $result['key'] = $result['editid'];
    }
    $result['title'] = trim(isset($_POST['title']) && strlen($_POST['title']) > 0 ? $_POST['title'] : (!isset($result['codeid']) && isset($_GET['title']) ? $_GET['title'] : ''));
    $result['post'] = isset($_POST['bbcode']) && strlen($_POST['bbcode']) > 0;
    $bbcode = trim($result['post'] ? $_POST['bbcode'] : (!isset($result['codeid']) && isset($_GET['bbcode']) ? $_GET['bbcode'] : ''));
    if( strlen($bbcode) > 0 && (strlen($bbcode) < 8 || (substr($bbcode, 0, 4) != '[map' && substr($bbcode, -6) != '[/map]')) )
        $bbcode = '[map]'.$bbcode.'[/map]';
    $result['bbcode'] = $bbcode;
    return $result;
}

?>
